src/lib/user_command_input/user_command_input.ml

```diff
@@ -0,0 +1,203 @@
+  end
+
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type t =
+        ( Common.Stable.V1.t
+        , User_command_payload.Body.Stable.V1.t )
+        User_command_payload.Poly.Stable.V1.t
+      [@@deriving sexp, to_yojson]
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  type t = Stable.Latest.t [@@deriving sexp, to_yojson]
+
+  let create ~fee ~nonce_opt ~valid_until ~memo ~body : t =
+    {common= Common.create ~fee ~nonce_opt ~valid_until ~memo; body}
+
+  let to_user_command_payload (t : t) ~nonce : User_command_payload.t =
+    { User_command_payload.Poly.common=
+        Common.to_user_command_common t.common ~nonce
+    ; body= t.body }
+end
+
+module Sign_choice = struct
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type t =
+        | Signature of Signature.Stable.V1.t
+        | Hd_index of Unsigned_extended.UInt32.Stable.V1.t
+        | Keypair of Keypair.Stable.V1.t
+      [@@deriving sexp, to_yojson]
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  type t = Stable.Latest.t =
+    | Signature of Signature.t
+    | Hd_index of Unsigned_extended.UInt32.t
+    | Keypair of Keypair.t
+  [@@deriving sexp, to_yojson]
+end
+
+[%%versioned
+module Stable = struct
+  module V1 = struct
+    type t =
+      ( Payload.Stable.V1.t
+      , Public_key.Compressed.Stable.V1.t
+      , Sign_choice.Stable.V1.t )
+      User_command.Poly.Stable.V1.t
+    [@@deriving sexp, to_yojson]
+
+    let to_latest = Fn.id
+  end
+end]
+
+type t = Stable.Latest.t [@@deriving sexp, to_yojson]
+
+let create ?(nonce_opt = None) ~fee ~valid_until ~memo ~body ~sender
+    ~sign_choice () : t =
+  let payload = Payload.create ~fee ~nonce_opt ~valid_until ~memo ~body in
+  {payload; sender; signature= sign_choice}
+
+let sign ~sender ~(user_command_payload : User_command_payload.t) = function
+  | Sign_choice.Signature signature ->
+      Option.value_map
+        ~default:(Deferred.return (Error "Invalid_signature"))
+        (User_command.create_with_signature_checked signature sender
+           user_command_payload)
+        ~f:Deferred.Result.return
+  | Keypair sender_kp ->
+      Deferred.Result.return (User_command.sign sender_kp user_command_payload)
+  | Hd_index hd_index ->
+      Secrets.Hardware_wallets.sign ~hd_index
+        ~public_key:(Public_key.decompress_exn sender)
+        ~user_command_payload
+
+let inferred_nonce ~f ~(sender : Public_key.Compressed.t) nonce_opt =
+  let open Result.Let_syntax in
+  let%bind txn_pool_or_account_nonce =
+    match Participating_state.active (f sender) |> Option.join with
+    | None ->
+        Error
+          "Couldn't infer nonce for transaction from specified `sender` since \
+           `sender` is not in the ledger or sent a transaction in transaction \
+           pool."
+    | Some nonce ->
+        Ok nonce
+  in
+  match nonce_opt with
+  | Some nonce ->
+      if Account_nonce.equal nonce txn_pool_or_account_nonce then Ok nonce
```

Comment:
This feels more like a failure of the user command input than of the nonce; might it make more sense to do this check in `to_user_command_common` instead?